### ISSUE TITLE
cv::split and nd-arrays

### ISSUE THREAD

**Issue by opencv-pushbot** [Itseez/opencv#5034]
Transferred from http://code.opencv.org/issues/4426
```
|| Amro _ on 2015-06-22 15:00
|| Priority: Normal
|| Affected: branch 'master' (3.0-dev)
|| Category: core
|| Tracker: Bug
|| Difficulty: 
|| PR: 
|| Platform: Any / Any
```

cv::split and nd-arrays
-----------
```
Using cv::split with an ND-array (dimensions not channels) drops dimensions higher than 2 in the result.

For instance:

<pre>
int dims[] = {5,4,3};
Mat m(3, dims, CV_8U); // or even: Mat m(3, dims, CV_8UC2)

vector<Mat> vm;
split(m, vm);          // all v[i].dims == 2 !
</pre>

Example code attached. Here is the output I get using OpenCV 3.0.0 and VS2013 x64:

<pre>
 --- cv::Mat ---
depth = 0
channels = 1
total = 60
dims = 3
size = {5, 4, 3, }
step = {12, 3, 1, }
data = 0x000000EF95525040

 --- vector<Mat> ---
vm.size() = 1

  vm[i]
depth = 0
channels = 1
total = 20
dims = 2
size = {5, 4, }
step = {4, 1, }
data = 0x000000EF97801530
</pre>
```

History
-------

**Comment by amroamroamro** [comment]
The attached example code from the bug report:

``` cpp
#include "opencv2/core.hpp"
#include <iostream>
#include <vector>
#include <string>
using namespace std;
using namespace cv;

void print_mat_info(const Mat& mat)
{
    cout << "depth = " << mat.depth() << endl;
    cout << "channels = " << mat.channels() << endl;
    cout << "total = " << mat.total() << endl;
    cout << "dims = " << mat.dims << endl;
    cout << "size = {";
    for (int i=0; i<mat.dims; i++)
        cout << mat.size[i] << ", ";
    cout << "}" << endl;
    cout << "step = {";
    for (int i=0; i<mat.dims; i++)
        cout << mat.step[i] << ", ";
    cout << "}" << endl;
    cout << "data = 0x" << reinterpret_cast<void*>(mat.data) << endl;
}

int main()
{
    // 3-dimensional array with 1-channel
    int dims[] = {5,4,3};
    Mat m(3, dims, CV_8UC1, Scalar::all(0));  // try: CV_8UC1, CV_8UC2

    // split across channels
    vector<Mat> vm;
    split(m, vm);

    cout << " --- cv::Mat --- " << endl;
    print_mat_info(m);

    cout << " --- vector<Mat> --- " << endl;
    cout << "vm.size() = " << vm.size() << endl;
    for (vector<Mat>::const_iterator it = vm.begin(); it != vm.end(); it++) {
        cout << "  vm[i]" << endl;
        print_mat_info(*it);
    }

    return 0;
}
```

**Comment by amroamroamro** [comment]
This bug is now fixed, thanks to @hoangviet1985 in #5644

@vpisarev 
Could someone please close the issue (both here and on code.opencv.org). I can't do it myself since it was auto transferred..
